### DISCUSSION TITLE
Bonders and deleters

### DISCUSSION THREAD

**Discussion by mathersrob** [lukasturcani/stk#293] [Q&A] [answered]
Hi,

I really like the idea behind bonders and deleters but find myself losing atoms in the process of creating a linear polymer with 2 building blocks containing carboxylic acids and alcohols. For example, below is some sample code from the stk Docs. I was hoping for esters but have lost the carbonyl. Any suggestions? Thanks!

alcohol_factory = stk.AlcoholFactory(
    # The index of the non-hydrogen atom connected to oxygen
    # is 0 in the functional group string
    bonders=(0, ),
    # The indices of the oxygen and hydrogen atoms in the
    # functional group string (see docstring) are 1 and 2, respectively.
    deleters=(1, ),
)
bb1 = stk.BuildingBlock(
    smiles='OCCCCO',
    functional_groups=(alcohol_factory, ),
)

carboxylic_acid_factory = stk.CarboxylicAcidFactory(
    # The index of the carbon atom in the functional
    # group string (see docstring) is 1.
    bonders=(1, ),
    # The index of the oxygen atom in the functional
    # group string (see docstring) is 2.
    deleters=(2, ),
)
bb2 = stk.BuildingBlock(
    smiles='OC(=O)CC(=O)O',
    functional_groups=(carboxylic_acid_factory, ),
)

polymer = stk.ConstructedMolecule(
    topology_graph=stk.polymer.Linear(
        building_blocks=(bb1, bb2),
        repeating_unit='AB',
        num_repeating_units=6,
        optimizer=stk.Collapser(scale_steps=False),
    ),
)

**Comment by lukasturcani** [comment, marked as answer]
Hi @mathersrob! Sorry for the delay in reply. If you would like to create esters try the following code:

```python
    # alcohol functional group string is [*][O][H]
    # see (https://stk.readthedocs.io/en/latest/stk.molecular.functional_groups.factories.alcohol_factory.html)
    alcohol_factory = stk.AlcoholFactory(
        # The index of the oxygen atom is 1 in the functional group string
        bonders=(1, ),
        # The index of the hydrogen atom in the
        # functional group string is 2 .
        deleters=(2, ),
    )
    bb1 = stk.BuildingBlock(
        smiles='OCCCCO',
        functional_groups=(alcohol_factory, ),
    )

    # carboxylic acid functional group string is [*][C](=[O])[O][H]
    # see (https://stk.readthedocs.io/en/latest/stk.molecular.functional_groups.factories.carboxylic_acid_factory.html)
    carboxylic_acid_factory = stk.CarboxylicAcidFactory(
        # The index of the carbon atom in the functional
        # group string is 1.
        bonders=(1, ),
        # The index of the non-carbonyl oxygen atom and
        # hydrogen atoms in the functional
        # group string are 3 and 4, respectively
        deleters=(3, 4),
    )
    bb2 = stk.BuildingBlock(
        smiles='OC(=O)CC(=O)O',
        functional_groups=(carboxylic_acid_factory, ),
    )

    polymer = stk.ConstructedMolecule(
        topology_graph=stk.polymer.Linear(
            building_blocks=(bb1, bb2),
            repeating_unit='AB',
            num_repeating_units=6,
            optimizer=stk.Collapser(scale_steps=False),
        ),
    )

```

I'll explain a bit about how / why this works. 

For the `AlcoholFactory` (https://stk.readthedocs.io/en/latest/stk.molecular.functional_groups.factories.alcohol_factory.html) the functional group string that is referred to in the docs is `[*][O][H]`.  This means the indices you pass to the `AlcoholFactory` refer to this string. So when using the `bonders` and  `deleters` arguments, 0 refers to the atom connected to the oxygen, 1 refers to  the oxygen and 2 refers to the hydrogen. Since for the ester reaction we want to add a bond to the oxygen atom on the alcohol and remove the hydrogen atom connected to it, we put 1 into the `bonders` list and `2` into the deleters list.

For the `CarboxylicAcidFactory` (https://stk.readthedocs.io/en/latest/stk.molecular.functional_groups.factories.carboxylic_acid_factory.html),  the functional group string that is referred to in the docs is `[*][C](=[O])[O][H]`. So the indices passed to `bonders` and `deleters` arguments of `CarboxylicAcidFactory` refer to this string. For the carboxylic acid, we want to bond to the oxygen so we put `1` as the bonder atom. We want to remove the alchol group, so we put 3 and 4 into the `deleters`paramter.

Hope this helps a bit, let me know if you run into any other issues and if you think the docs could be improved in some way.

**Reply by mathersrob** [reply]
Hi @lukasturcani,

Thanks for the answer. Very helpful! 

I now see that the numbering for [ * ][O][H] and [ * ][C](=[O])[O][H] goes L to R. This was not initially obvious. Since oxygen has two atoms connected to it, I got a little confused by your statement "0 refers to the atom connected to the oxygen."  

best,

Rob

**Reply by lukasturcani** [reply]
Ah, that makes sense! Thanks for the feedback!

**Reply by lukasturcani** [reply]
I've added an issue to track this, https://github.com/lukasturcani/stk/issues/294
